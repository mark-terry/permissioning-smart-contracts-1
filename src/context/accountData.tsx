import React, {
    createContext,
    useContext,
    useEffect,
    useState,
    useMemo
} from "react";
import { drizzleReactHooks } from "drizzle-react";

type Account = {address: string}

type ContextType = {
    accountWhitelist?: Account[],
    setAccountWhitelist?: (account: Account[]) => void
}

const AccountDataContext = createContext<ContextType>({})

/**
 * Provider for the data context that contains the whitelist
 * @param {Object} props Props given to the DataProvider
 * @return The provider with the following value:
 *  - nodeWhitelist: list of whiteliist enode from Node Rules contract
 *  - setNodeWhitelist: setter for the whitelist state
 */
export const AccountDataProvider: React.FC = (props: React.Props<{}>) => {
    const [accountWhitelist, setAccountWhitelist] = useState<Account[]>([])
    const value = useMemo(() => ({ accountWhitelist, setAccountWhitelist }), [
        accountWhitelist,
        setAccountWhitelist
    ])
    return <AccountDataContext.Provider value={value} {...props} />
}

/**
 * Fetch the appropriate data on chain and synchronize with it
 * @return {Object} Contains data of interest:
 *  - admins: list of admin address from Admin contract,
 *  - dataReady: true if isReadOnly, whitelist and admins are correctly fetched,
 *  false otherwise
 *  - userAddress: Address of the user
 *  - isAdmin: true if address of the user is includes in the admin list
 *  - node: Object containing node relevant data
 *    - isReadOnly: Rules contract is lock or unlock,
 *    - whitelist: list of whitelist enode from Rules contract,
 */
export const useAccountData = () => {
    const context = useContext(AccountDataContext)

    if (!context) {
        throw new Error("useAccountData must be used within a AccountDataProvider.")
    }

    const { accountWhitelist, setAccountWhitelist } = context
    const { drizzle, useCacheCall } = drizzleReactHooks.useDrizzle()
    const accountIsReadOnly: boolean = useCacheCall("AccountRules", "isReadOnly")
    const accountWhitelistSize: number = useCacheCall("AccountRules", "getSize")
    const { getByIndex: getAccountByIndex } = drizzle.contracts.AccountRules.methods
    const { userAddress } = drizzleReactHooks.useDrizzleState((drizzleState: any) => ({
        userAddress: drizzleState.accounts[0]
    }));

    useEffect(() => {
        const promises = [];
        for (let index = 0; index < accountWhitelistSize; index++) {
            promises.push(getAccountByIndex(index).call());
        }
        Promise.all(promises).then(responses => {
            const updatedAccountWhitelist = responses.map((address: string) => ({address}));
            setAccountWhitelist!(updatedAccountWhitelist);
        })
    }, [accountWhitelistSize, setAccountWhitelist, getAccountByIndex])

    const dataReady = useMemo(
        () =>
            typeof accountIsReadOnly === "boolean" &&
            Array.isArray(accountWhitelist),
        [accountIsReadOnly, accountWhitelist]
    )

    const formattedAccountWhitelist = useMemo(() => {
        return accountWhitelist
            ? accountWhitelist
                  .map(account => ({ ...account, status: "active" }))
                  .reverse()
            : undefined
    }, [accountWhitelist])

    return {
        userAddress,
        dataReady,
        whitelist: formattedAccountWhitelist,
        isReadOnly: accountIsReadOnly
    }
}