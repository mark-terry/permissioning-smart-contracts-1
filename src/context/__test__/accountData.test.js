// Libs
import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
// Components
import { AccountDataProvider } from "../accountData";

describe("<AccountDataProvider />", () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <AccountDataProvider>
                <div className="test" />
            </AccountDataProvider>
        );
    });

    it("renders children when passed in", () => {
        expect(wrapper.contains(<div className="test" />)).toEqual(true);
    });

    it("matches snapshot", () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
