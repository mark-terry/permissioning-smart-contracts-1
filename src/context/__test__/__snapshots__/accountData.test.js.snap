// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountDataProvider /> matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "accountWhitelist": Array [],
      "setAccountWhitelist": [Function],
    }
  }
>
  <div
    className="test"
  />
</ContextProvider>
`;
